File: apps/agent-worker/worker/scenario_engine/schematic_strategy.py
"""Schematic / generic HCP MLR PRB cadence (parity with schematic_strategy.ts)."""

from __future__ import annotations

from typing import Any, Literal, Mapping, TypedDict

from worker.scenario_engine.date_calendar import add_calendar_days_utc, parse_iso_date_utc
from worker.scenario_engine.prb_weekday_anchors import (
    first_working_monday_on_or_after,
    resolve_prb_anchor_day,
    second_working_wednesday_after_monday_submit,
)
from worker.scenario_engine.scenario_linear_shared import (
    linear_allow_non_working_for,
    linear_clamp_days,
    linear_prb_step_allow_non_working,
    linear_step_payload,
)
from worker.scenario_engine.timing_profiles import uses_schematic_mlr_prb_cadence
from worker.scenario_engine.working_days import (
    HolidaySet,
    add_working_days_utc,
    inclusive_working_day_span,
    next_working_day,
    previous_working_day_on_or_before,
)

MLR_PRB_IDS = frozenset({
    'submit_prb1',
    'prb1_review',
    'submit_prb2',
    'prb2_review',
    'submit_prb3',
    'prb3_review',
})


class SchematicMlrPrbRefs(TypedDict):
    email_prb1_submit_monday: str | None
    email_prb1_submit_end: str | None
    email_prb2_submit_monday: str | None
    email_prb2_submit_end: str | None
    email_prb3_submit_monday: str | None
    email_prb3_submit_end: str | None


def schematic_mlr_apply_pinned_prefix_prb_state(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    pin: Mapping[str, Any],
    refs: SchematicMlrPrbRefs,
) -> None:
    rid = str(row['id'])
    if not uses_schematic_mlr_prb_cadence(timing_profile) or allow or rid not in MLR_PRB_IDS:
        return
    if rid == 'submit_prb1':
        refs['email_prb1_submit_monday'] = str(pin['start_date'])
        refs['email_prb1_submit_end'] = str(pin['end_date'])
    elif rid == 'submit_prb2':
        refs['email_prb2_submit_monday'] = str(pin['start_date'])
        refs['email_prb2_submit_end'] = str(pin['end_date'])
    elif rid == 'submit_prb3':
        refs['email_prb3_submit_monday'] = str(pin['start_date'])
        refs['email_prb3_submit_end'] = str(pin['end_date'])


def schematic_mlr_adjust_effective_for_development_dev_gap(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    holidays: HolidaySet,
    allow_map: dict[str, bool] | None,
    mod_sum: int,
    effective: int,
    refs: SchematicMlrPrbRefs,
) -> int | str:
    rid = str(row['id'])
    dev_prb_row = rid in ('development_prb1', 'development_prb2', 'development_prb3')
    use_email_prb_dev_gap = uses_schematic_mlr_prb_cadence(timing_profile) and not allow and dev_prb_row
    if not use_email_prb_dev_gap:
        return effective

    submit_mon_a: str | None = None
    submit_end_a: str | None = None
    review_phase_e: Literal['prb1_review', 'prb2_review', 'prb3_review'] = 'prb1_review'
    err_early = 'PRB1 development scheduled before PRB1 submit (internal error).'
    if rid == 'development_prb1':
        submit_mon_a = refs['email_prb1_submit_monday']
        submit_end_a = refs['email_prb1_submit_end']
        review_phase_e = 'prb1_review'
        err_early = 'PRB1 development scheduled before PRB1 submit (internal error).'
    elif rid == 'development_prb2':
        submit_mon_a = refs['email_prb2_submit_monday']
        submit_end_a = refs['email_prb2_submit_end']
        review_phase_e = 'prb2_review'
        err_early = 'PRB2 development scheduled before PRB2 submit (internal error).'
    else:
        submit_mon_a = refs['email_prb3_submit_monday']
        submit_end_a = refs['email_prb3_submit_end']
        review_phase_e = 'prb3_review'
        err_early = 'PRB3 development scheduled before PRB3 submit (internal error).'
    if not submit_mon_a or not submit_end_a:
        return err_early
    ideal_wed = second_working_wednesday_after_monday_submit(submit_mon_a, holidays)
    review_allow = linear_allow_non_working_for(review_phase_e, allow_map)
    review_resolved = resolve_prb_anchor_day(ideal_wed, holidays, review_allow)
    gap_start = next_working_day(add_calendar_days_utc(submit_end_a, 1), holidays)
    gap_end = previous_working_day_on_or_before(
        add_calendar_days_utc(str(review_resolved['iso']), -1), holidays
    )
    if parse_iso_date_utc(gap_start) > parse_iso_date_utc(gap_end):
        gap_start = gap_end
    raw_span = max(1, inclusive_working_day_span(gap_start, gap_end, holidays))
    bumped = linear_clamp_days(raw_span + mod_sum, row)
    return min(bumped, raw_span)


def schematic_mlr_try_place_prb_row(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    cursor_cal: str,
    holidays: HolidaySet,
    allow_map: dict[str, bool] | None,
    effective: int,
    note_out: str,
    refs: SchematicMlrPrbRefs,
    steps_out: list[dict[str, Any]],
) -> tuple[Literal['placed'], str] | tuple[Literal['skip'], None] | tuple[Literal['err'], str]:
    rid = str(row['id'])
    use_email_prb_cadence = uses_schematic_mlr_prb_cadence(timing_profile) and not allow and rid in MLR_PRB_IDS
    if not use_email_prb_cadence:
        return ('skip', None)

    phase_allow = linear_allow_non_working_for(rid, allow_map)

    if rid == 'submit_prb1':
        min_start = next_working_day(cursor_cal, holidays)
        merged = min_start
        mon_raw = first_working_monday_on_or_after(merged, holidays)
        resolved = resolve_prb_anchor_day(mon_raw, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        refs['email_prb1_submit_monday'] = start
        refs['email_prb1_submit_end'] = end
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'prb1_review':
        if not refs['email_prb1_submit_monday']:
            return ('err', 'PRB1 review scheduled before PRB1 submit (internal error).')
        ideal_wed = second_working_wednesday_after_monday_submit(refs['email_prb1_submit_monday'], holidays)
        resolved = resolve_prb_anchor_day(ideal_wed, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'submit_prb2':
        min_start = next_working_day(cursor_cal, holidays)
        merged = min_start
        mon_raw = first_working_monday_on_or_after(merged, holidays)
        resolved = resolve_prb_anchor_day(mon_raw, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        refs['email_prb2_submit_monday'] = start
        refs['email_prb2_submit_end'] = end
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'prb2_review':
        if not refs['email_prb2_submit_monday']:
            return ('err', 'PRB2 review scheduled before PRB2 submit (internal error).')
        ideal_wed = second_working_wednesday_after_monday_submit(refs['email_prb2_submit_monday'], holidays)
        resolved = resolve_prb_anchor_day(ideal_wed, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'submit_prb3':
        min_start = next_working_day(cursor_cal, holidays)
        merged = min_start
        mon_raw = first_working_monday_on_or_after(merged, holidays)
        resolved = resolve_prb_anchor_day(mon_raw, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        refs['email_prb3_submit_monday'] = start
        refs['email_prb3_submit_end'] = end
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'prb3_review':
        if not refs['email_prb3_submit_monday']:
            return ('err', 'PRB3 review scheduled before PRB3 submit (internal error).')
        ideal_wed = second_working_wednesday_after_monday_submit(refs['email_prb3_submit_monday'], holidays)
        resolved = resolve_prb_anchor_day(ideal_wed, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    return ('skip', None)


def schematic_mlr_try_place_development_dev_gap_row(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    holidays: HolidaySet,
    allow_map: dict[str, bool] | None,
    effective: int,
    note_out: str,
    refs: SchematicMlrPrbRefs,
    steps_out: list[dict[str, Any]],
) -> tuple[Literal['placed'], str] | tuple[Literal['skip'], None] | tuple[Literal['err'], str]:
    rid = str(row['id'])
    dev_prb_row = rid in ('development_prb1', 'development_prb2', 'development_prb3')
    use_email_prb_dev_gap = uses_schematic_mlr_prb_cadence(timing_profile) and not allow and dev_prb_row
    if not use_email_prb_dev_gap:
        return ('skip', None)

    submit_mon = (
        refs['email_prb1_submit_monday']
        if rid == 'development_prb1'
        else refs['email_prb2_submit_monday']
        if rid == 'development_prb2'
        else refs['email_prb3_submit_monday']
    )
    submit_end = (
        refs['email_prb1_submit_end']
        if rid == 'development_prb1'
        else refs['email_prb2_submit_end']
        if rid == 'development_prb2'
        else refs['email_prb3_submit_end']
    )
    review_phase = (
        'prb1_review' if rid == 'development_prb1' else 'prb2_review' if rid == 'development_prb2' else 'prb3_review'
    )
    if submit_mon is None or submit_end is None:
        return ('err', 'PRB development gap missing submit anchor (internal error).')
    ideal_wed = second_working_wednesday_after_monday_submit(submit_mon, holidays)
    review_resolved = resolve_prb_anchor_day(
        ideal_wed,
        holidays,
        linear_allow_non_working_for(review_phase, allow_map),
    )
    gap_start = next_working_day(add_calendar_days_utc(submit_end, 1), holidays)
    gap_end = previous_working_day_on_or_before(
        add_calendar_days_utc(str(review_resolved['iso']), -1), holidays
    )
    if parse_iso_date_utc(gap_start) > parse_iso_date_utc(gap_end):
        gap_start = gap_end
    start = gap_start
    end = add_working_days_utc(start, effective - 1, holidays)
    steps_out.append(linear_step_payload(str(row['label']), start, end, note_out, False))
    return ('placed', add_calendar_days_utc(end, 1))

File: apps/web-dashboard/lib/scenarioPlanner/linear/schematic_strategy.ts
/**
 * Schematic / generic HCP MLR PRB cadence: Monday submit + second working Wednesday review (`prb_cadence: email_ml_r` in config).
 */

import type { HalTimelineStep } from '../../halScenario';
import { addCalendarDaysUTC, parseIsoDateUTC } from '../dateCalendar';
import {
  firstWorkingMondayOnOrAfter,
  resolvePrbAnchorDay,
  secondWorkingWednesdayAfterMondaySubmit,
} from '../prbWeekdayAnchors';
import type { ScenarioTactic } from '../tactics';
import { usesSchematicMlrPrbCadence } from '../timingProfiles';
import {
  addWorkingDaysUTC,
  inclusiveWorkingDaySpan,
  nextWorkingDay,
  previousWorkingDayOnOrBefore,
  type HolidaySet,
} from '../workingDays';
import type { ScenarioStepDef } from './types';
import { linearAllowNonWorkingFor, linearClampDays, linearPrbStepAllowNonWorking, linearStepPayload } from './scenario_linear_shared';

const MLR_PRB_IDS = new Set([
  'submit_prb1',
  'prb1_review',
  'submit_prb2',
  'prb2_review',
  'submit_prb3',
  'prb3_review',
]);

export type SchematicMlrPrbRefs = {
  emailPrb1SubmitMonday: string | null;
  emailPrb1SubmitEnd: string | null;
  emailPrb2SubmitMonday: string | null;
  emailPrb2SubmitEnd: string | null;
  emailPrb3SubmitMonday: string | null;
  emailPrb3SubmitEnd: string | null;
};

export function schematicMlrApplyPinnedPrefixPrbState(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  pin: HalTimelineStep,
  refs: SchematicMlrPrbRefs,
): void {
  if (!usesSchematicMlrPrbCadence(timingProfile) || allow || !MLR_PRB_IDS.has(row.id)) return;
  if (row.id === 'submit_prb1') {
    refs.emailPrb1SubmitMonday = pin.start_date;
    refs.emailPrb1SubmitEnd = pin.end_date;
  } else if (row.id === 'submit_prb2') {
    refs.emailPrb2SubmitMonday = pin.start_date;
    refs.emailPrb2SubmitEnd = pin.end_date;
  } else if (row.id === 'submit_prb3') {
    refs.emailPrb3SubmitMonday = pin.start_date;
    refs.emailPrb3SubmitEnd = pin.end_date;
  }
}

export function schematicMlrAdjustEffectiveForDevelopmentDevGap(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  holidays: HolidaySet,
  allowMap: Readonly<Record<string, boolean>> | undefined,
  modSum: number,
  effective: number,
  refs: SchematicMlrPrbRefs,
): number | { ok: false; error: string } {
  const devPrbRow =
    row.id === 'development_prb1' || row.id === 'development_prb2' || row.id === 'development_prb3';
  const useEmailPrbDevGap = usesSchematicMlrPrbCadence(timingProfile) && !allow && devPrbRow;
  if (!useEmailPrbDevGap) return effective;

  let submitMonday: string | null = null;
  let submitEnd: string | null = null;
  let reviewPhaseId: 'prb1_review' | 'prb2_review' | 'prb3_review';
  let errEarly: string | null = null;
  if (row.id === 'development_prb1') {
    submitMonday = refs.emailPrb1SubmitMonday;
    submitEnd = refs.emailPrb1SubmitEnd;
    reviewPhaseId = 'prb1_review';
    errEarly = 'PRB1 development scheduled before PRB1 submit (internal error).';
  } else if (row.id === 'development_prb2') {
    submitMonday = refs.emailPrb2SubmitMonday;
    submitEnd = refs.emailPrb2SubmitEnd;
    reviewPhaseId = 'prb2_review';
    errEarly = 'PRB2 development scheduled before PRB2 submit (internal error).';
  } else {
    submitMonday = refs.emailPrb3SubmitMonday;
    submitEnd = refs.emailPrb3SubmitEnd;
    reviewPhaseId = 'prb3_review';
    errEarly = 'PRB3 development scheduled before PRB3 submit (internal error).';
  }
  if (!submitMonday || !submitEnd) {
    return { ok: false, error: errEarly! };
  }
  const idealWed = secondWorkingWednesdayAfterMondaySubmit(submitMonday, holidays);
  const reviewAllow = linearAllowNonWorkingFor(reviewPhaseId, allowMap);
  const reviewResolved = resolvePrbAnchorDay(idealWed, holidays, reviewAllow);
  let gapStart = nextWorkingDay(addCalendarDaysUTC(submitEnd, 1), holidays);
  const gapEnd = previousWorkingDayOnOrBefore(addCalendarDaysUTC(reviewResolved.iso, -1), holidays);
  if (parseIsoDateUTC(gapStart).getTime() > parseIsoDateUTC(gapEnd).getTime()) {
    gapStart = gapEnd;
  }
  const rawSpan = Math.max(1, inclusiveWorkingDaySpan(gapStart, gapEnd, holidays));
  const bumped = linearClampDays(rawSpan + modSum, row);
  return Math.min(bumped, rawSpan);
}

export function schematicMlrTryPlacePrbRow(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  cursorCal: string,
  holidays: HolidaySet,
  allowMap: Readonly<Record<string, boolean>> | undefined,
  effective: number,
  noteOut: string,
  refs: SchematicMlrPrbRefs,
  steps: HalTimelineStep[],
): { placed: true; nextCursorCal: string } | { placed: false } | { ok: false; error: string } {
  const useEmailPrbCadence =
    usesSchematicMlrPrbCadence(timingProfile) &&
    !allow &&
    MLR_PRB_IDS.has(row.id);
  if (!useEmailPrbCadence) return { placed: false };

  const phaseAllow = linearAllowNonWorkingFor(row.id, allowMap);

  if (row.id === 'submit_prb1') {
    const minStart = nextWorkingDay(cursorCal, holidays);
    const merged = minStart;
    const monRaw = firstWorkingMondayOnOrAfter(merged, holidays);
    const resolved = resolvePrbAnchorDay(monRaw, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    refs.emailPrb1SubmitMonday = start;
    refs.emailPrb1SubmitEnd = end;
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'prb1_review') {
    if (!refs.emailPrb1SubmitMonday) {
      return { ok: false, error: 'PRB1 review scheduled before PRB1 submit (internal error).' };
    }
    const idealWed = secondWorkingWednesdayAfterMondaySubmit(refs.emailPrb1SubmitMonday, holidays);
    const resolved = resolvePrbAnchorDay(idealWed, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'submit_prb2') {
    const minStart = nextWorkingDay(cursorCal, holidays);
    const merged = minStart;
    const monRaw = firstWorkingMondayOnOrAfter(merged, holidays);
    const resolved = resolvePrbAnchorDay(monRaw, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    refs.emailPrb2SubmitMonday = start;
    refs.emailPrb2SubmitEnd = end;
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'prb2_review') {
    if (!refs.emailPrb2SubmitMonday) {
      return { ok: false, error: 'PRB2 review scheduled before PRB2 submit (internal error).' };
    }
    const idealWed = secondWorkingWednesdayAfterMondaySubmit(refs.emailPrb2SubmitMonday, holidays);
    const resolved = resolvePrbAnchorDay(idealWed, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'submit_prb3') {
    const minStart = nextWorkingDay(cursorCal, holidays);
    const merged = minStart;
    const monRaw = firstWorkingMondayOnOrAfter(merged, holidays);
    const resolved = resolvePrbAnchorDay(monRaw, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    refs.emailPrb3SubmitMonday = start;
    refs.emailPrb3SubmitEnd = end;
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'prb3_review') {
    if (!refs.emailPrb3SubmitMonday) {
      return { ok: false, error: 'PRB3 review scheduled before PRB3 submit (internal error).' };
    }
    const idealWed = secondWorkingWednesdayAfterMondaySubmit(refs.emailPrb3SubmitMonday, holidays);
    const resolved = resolvePrbAnchorDay(idealWed, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  return { placed: false };
}

export function schematicMlrTryPlaceDevelopmentDevGapRow(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  holidays: HolidaySet,
  allowMap: Readonly<Record<string, boolean>> | undefined,
  effective: number,
  noteOut: string,
  refs: SchematicMlrPrbRefs,
  steps: HalTimelineStep[],
): { placed: true; nextCursorCal: string } | { placed: false } {
  const devPrbRow =
    row.id === 'development_prb1' || row.id === 'development_prb2' || row.id === 'development_prb3';
  const useEmailPrbDevGap = usesSchematicMlrPrbCadence(timingProfile) && !allow && devPrbRow;
  if (!useEmailPrbDevGap) return { placed: false };

  const submitMonday =
    row.id === 'development_prb1'
      ? refs.emailPrb1SubmitMonday!
      : row.id === 'development_prb2'
        ? refs.emailPrb2SubmitMonday!
        : refs.emailPrb3SubmitMonday!;
  const submitEnd =
    row.id === 'development_prb1'
      ? refs.emailPrb1SubmitEnd!
      : row.id === 'development_prb2'
        ? refs.emailPrb2SubmitEnd!
        : refs.emailPrb3SubmitEnd!;
  const reviewPhaseId =
    row.id === 'development_prb1' ? 'prb1_review' : row.id === 'development_prb2' ? 'prb2_review' : 'prb3_review';
  const idealWed = secondWorkingWednesdayAfterMondaySubmit(submitMonday, holidays);
  const reviewResolved = resolvePrbAnchorDay(
    idealWed,
    holidays,
    linearAllowNonWorkingFor(reviewPhaseId, allowMap),
  );
  let gapStart = nextWorkingDay(addCalendarDaysUTC(submitEnd, 1), holidays);
  const gapEnd = previousWorkingDayOnOrBefore(addCalendarDaysUTC(reviewResolved.iso, -1), holidays);
  if (parseIsoDateUTC(gapStart).getTime() > parseIsoDateUTC(gapEnd).getTime()) {
    gapStart = gapEnd;
  }
  const start = gapStart;
  const end = addWorkingDaysUTC(start, effective - 1, holidays);
  steps.push(linearStepPayload(row.label, start, end, noteOut, false));
  return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
}
